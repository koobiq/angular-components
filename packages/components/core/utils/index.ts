export * from './dom';
export * from './public-api';
