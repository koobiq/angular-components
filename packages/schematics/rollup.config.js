const typescript = require('@rollup/plugin-typescript');
const replace = require('@rollup/plugin-replace');

const path = require('path');
const { promises: fs } = require('fs');

const pkg = require('../../package.json');
const { getMigrations } = require('./src/utils/migrations');

const version = (str) => JSON.stringify(str.startsWith('^') ? str : '^' + str);

const distDir = path.join(__dirname, 'dist');

const clean = () => ({
    async buildStart() {
        await fs.rm(distDir, { force: true, recursive: true });
    }
});

module.exports = [
    {
        output: {
            dir: distDir,
            format: 'cjs',
            exports: 'named'
        },
        input: {
            'ng-add/index': path.join(__dirname, 'src/ng-add/index.ts'),
            'utils/package-config': path.join(__dirname, 'src/utils/package-config.ts'),
            'utils/messages': path.join(__dirname, 'src/utils/messages.ts'),
            ...getMigrations().reduce((res, cur) => {
                res[`migrations/${cur}/index`] = path.join(__dirname, `src/migrations/${cur}/index.ts`);
                res[`migrations/${cur}/data`] = path.join(__dirname, `src/migrations/${cur}/data.ts`);
                return res;
            }, {})
            //'ng-add/setup-project': path.join(__dirname, 'ng-add/setup-project.ts'),
        },
        external: (dependency) =>
            !(dependency.startsWith('./') || dependency.startsWith('../') || dependency.startsWith(__dirname)),
        plugins: [
            clean(),
            replace({
                preventAssignment: true,
                'VERSIONS.ANGULAR_CDK': version(pkg.dependencies['@angular/cdk']),
                'VERSIONS.KOOBIQ_CDK': version(pkg.version),
                'VERSIONS.KOOBIQ_ANGULAR_LUXON_ADAPTER': version(pkg.version),
                'VERSIONS.KOOBIQ_DATE_FORMATTER': version(pkg.dependencies['@koobiq/date-formatter']),
                'VERSIONS.KOOBIQ_DATE_ADAPTER': version(pkg.dependencies['@koobiq/date-adapter']),
                'VERSIONS.KOOBIQ_TOKENS_BUILDER': version(pkg.devDependencies['@koobiq/tokens-builder']),
                'VERSIONS.KOOBIQ_DESIGN_TOKENS': version(pkg.devDependencies['@koobiq/design-tokens']),
                'VERSIONS.KOOBIQ_ICONS': version(pkg.dependencies['@koobiq/icons']),
                'VERSIONS.MESSAGEFORMAT_CORE': version(pkg.devDependencies['@messageformat/core']),
                'VERSIONS.LUXON': version(pkg.devDependencies.luxon),
                'VERSIONS.NGX_HIGHLIGHTJS': version(pkg.devDependencies['ngx-highlightjs'])
            }),
            typescript({
                tsconfig: path.join(__dirname, 'tsconfig.rollup.json')
            })

        ]
    }
];
