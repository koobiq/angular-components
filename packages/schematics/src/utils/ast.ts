/** WARNING:
 * this is copy of https://github.com/angular/angular/blob/main/packages/compiler/src/ml_parser/ast.ts
 * since @angular/compiler and @angular-devkit/schematics
 * are not compatible with each other */
export function visitAll(visitor, nodes): any[] {
    const result: any[] = [];

    const visit = visitor.visit
        ? (ast: Element) => visitor.visit!(ast) || ast.visit(visitor)
        : (ast: Element) => ast.visit(visitor);
    nodes.forEach((ast) => {
        visit(ast);
    });
    return result;
}

export interface ParseSourceFile {
    content: string;
    url: string;
}

export interface ParseLocation {
    file: ParseSourceFile;
    offset: number;
    line: number;
    col: number;
}

export interface Element {
    name: string;
    attrs: Attribute[];
    children: Element[];
    /**
     * actual code from angular/compiler:
     * override visit(visitor: Visitor, context: any): any {
     *   return visitor.visitElement(this, context);
     * }
     */
    visit(visitor: any, context?: any): any;
}

export interface Attribute {
    name: string;
    value: string;
    valueTokens: any[];
    keySpan: any;
    valueSpan: { start: ParseLocation; end: ParseLocation; fullStart: any; details: any };
}

export interface Block {
    name: string;
    children: Element[];
    visit(visitor: Visitor, context: any): any;
}

export interface Visitor {
    // Returning a truthy value from `visit()` will prevent `visitAll()` from the call to the typed
    // method and result returned will become the result included in `visitAll()`s result array.
    visit?(node: Node, context: any): any;

    visitElement(element: Element, context: any): any;
    visitAttribute(attribute: Attribute, context: any): any;
    visitText(text: Text, context: any): any;
    visitComment(comment: Comment, context: any): any;
    visitExpansion(expansion: any, context: any): any;
    visitExpansionCase(expansionCase: any, context: any): any;
    visitBlock(block: any, context: any): any;
    visitBlockParameter(parameter: any, context: any): any;
    visitLetDeclaration(decl: any, context: any): any;
}

// attr can be naked or in [ ]
// attr.value can be {{ value }} , 'value'
export function getSimpleAttributeName(attrName: string): string {
    return attrName.replace(/[\[\]]/g, '');
}

export function getSimpleAttributeValue(attrValue: string): string {
    return attrValue.replace(/[{}']/g, '');
}
