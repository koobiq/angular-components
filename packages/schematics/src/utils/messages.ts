export function noProject(project: string) {
    return `Unable to find project '${project}' in the workspace`;
}

export const logMessage = (logger, content) => {
    logger.warn('-------------------------');
    logger.warn(content.join('\n'));
    logger.warn('-------------------------');
};
