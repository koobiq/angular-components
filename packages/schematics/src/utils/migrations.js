const { readdirSync } = require('fs');
const { resolve } = require('path');

const getMigrations = () =>
    readdirSync(resolve(__dirname, '../migrations'), { withFileTypes: true })
        .filter((file) => file.isDirectory())
        .map((dir) => dir.name);

module.exports = {
    getMigrations
};
