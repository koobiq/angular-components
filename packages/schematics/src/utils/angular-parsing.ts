import { SchematicContext, Tree } from '@angular-devkit/schematics';
import { relative } from 'path';
import ts from 'typescript';
import { getSimpleAttributeName, getSimpleAttributeValue, visitAll } from './ast';
import {
    AnalyzedFile,
    ElementCollector,
    forEachClass,
    MigrationData,
    parseTemplate,
    TransformTemplateAttributesResult
} from './typescript';

/**
 * Update separate html file
 * @param tree
 * @param filePaths
 * @param context
 * @param migrationData
 */
export async function migrateTemplate(
    tree: Tree,
    filePaths: string[],
    context: SchematicContext,
    migrationData: MigrationData
) {
    for (const templatePath of filePaths) {
        const parsedFilePath = `.${templatePath}`;
        let res: Promise<TransformTemplateAttributesResult> = Promise.resolve({
            fileContent: '',
            changed: false,
            errors: []
        });

        const template = tree.read(parsedFilePath)?.toString();

        if (!template) {
            context.logger.error(`Cannot read "${parsedFilePath}" because it does not exist.`);
            continue;
        }

        res = transformTemplateAttributes(template, parsedFilePath, migrationData);

        const { fileContent, changed, errors } = await res;

        if (changed) {
            tree.overwrite(templatePath, fileContent);
        }

        if (errors.length > 0) {
            context.logger.error(errors.map(({ error }) => error.toString()).join('\n'));
        }
    }
}

/**
 * Update typescript file if classes with @Component decorator and `template` property exists
 * @param tree
 * @param sourceFiles
 * @param basePath
 * @param context
 * @param migrationData
 */
export async function migrateTs(
    tree: Tree,
    sourceFiles: ts.SourceFile[],
    basePath: string,
    context: SchematicContext,
    migrationData: MigrationData
) {
    // run and track changed components
    const analysis = new Map<string, AnalyzedFile>();

    for (const sourceFile of sourceFiles) {
        forEachClass(sourceFile, (node) => {
            analyzeDecorators(node, sourceFile, analysis);
        });
    }

    // update source file
    for (const path of analysis.keys()) {
        const file = analysis.get(path)!;
        const ranges = file.getSortedRanges();
        const relativePath = relative(basePath, path);
        const content = tree.readText(relativePath);
        const update = tree.beginUpdate(relativePath);

        for (const { start, end } of ranges) {
            const template = content.slice(start, end);
            const length = (end ?? content.length) - start;

            const { fileContent, changed, errors } = await transformTemplateAttributes(
                template,
                relativePath,
                migrationData
            );

            if (changed) {
                update.remove(start, length);
                update.insertLeft(start, fileContent);
            }

            if (errors.length > 0) {
                context.logger.error(errors.map(({ error }) => error.toString()).join('\n'));
            }
        }

        tree.commitUpdate(update);
    }
}

/**
 * Analyzes the decorators of a given class declaration to find and save
 * inline `template` metadata within `@Component` decorators.
 */
export function analyzeDecorators(
    node: ts.ClassDeclaration,
    sourceFile: ts.SourceFile,
    analyzedFiles: Map<string, AnalyzedFile>
) {
    const decorator = ts.getDecorators(node)?.find((dec) => {
        return (
            ts.isCallExpression(dec.expression) &&
            ts.isIdentifier(dec.expression.expression) &&
            dec.expression?.expression?.text === 'Component'
        );
    }) as (ts.Decorator & { expression: ts.CallExpression }) | undefined;

    const metadata =
        decorator &&
        decorator.expression.arguments.length > 0 &&
        ts.isObjectLiteralExpression(decorator.expression.arguments[0])
            ? decorator.expression.arguments[0]
            : null;

    // Exit early if there's no metadata or it's not an object literal
    if (!metadata) {
        return;
    }

    for (const prop of metadata.properties) {
        // All the properties we care about should have static
        // names and be initialized to a static string.
        if (!ts.isPropertyAssignment(prop) || (!ts.isIdentifier(prop.name) && !ts.isStringLiteralLike(prop.name))) {
            continue;
        }

        // check decorator 'template' property, check for text content parse
        if (prop.name.text === 'template' && ts.isStringLiteralLike(prop.initializer) && prop.initializer.text) {
            // +1/-1 to exclude the opening/closing characters from the range.
            AnalyzedFile.addRange(sourceFile.fileName, sourceFile, analyzedFiles, {
                start: prop.initializer.getStart(sourceFile) + 1,
                end: prop.initializer.getEnd() - 1,
                node: prop,
                type: 'template',
                remove: true
            });
        }
    }
}

/**
 * Migrates attribute values in an Angular template based on predefined cases.
 */
async function transformTemplateAttributes(
    template: string,
    fileName: string,
    migrationData: MigrationData
): Promise<TransformTemplateAttributesResult> {
    const parsed = await parseTemplate(template);
    let updatedTemplate = template;
    // extra text offset as result of text replacement;
    let offset = 0;

    if (parsed.tree !== undefined) {
        const visitor = new ElementCollector(migrationData, fileName);

        visitAll(visitor, parsed.tree.rootNodes);

        for (const el of visitor.elementsToMigrate) {
            const migrationAttr = el.attrs.find(
                (attr) => getSimpleAttributeName(attr.name) === migrationData.attrs.key.from
            );
            let updatedAttrValue: string | undefined;

            if (migrationAttr) {
                const cleanAttrValue = getSimpleAttributeValue(migrationAttr.value);
                const replacement = migrationData.attrs.value.replacements.find(({ from }) => from === cleanAttrValue);

                if (replacement) {
                    updatedAttrValue = replacement.to;
                } else {
                    console.warn(
                        `Element is using dynamic value. Check the code and change value on your own.${
                            (fileName && ' File: ' + fileName) || ''
                        }`
                    );
                    continue;
                }

                updatedTemplate =
                    updatedTemplate.slice(0, migrationAttr.keySpan.start.offset - offset) +
                    `${migrationData.attrs.key.to}="${updatedAttrValue}"` +
                    updatedTemplate.slice(migrationAttr.valueSpan.end.offset + 1 - offset, updatedTemplate.length);

                offset += migrationAttr.name.length - migrationData.attrs.key.to.length;
                offset += migrationAttr.value.length - updatedAttrValue.length;
            }
        }

        return { fileContent: updatedTemplate, changed: visitor.elementsToMigrate.length > 0, errors: parsed.errors };
    }

    return { fileContent: updatedTemplate, changed: undefined, errors: parsed.errors };
}
