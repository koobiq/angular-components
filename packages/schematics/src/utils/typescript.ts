import ts from 'typescript';
import { Attribute, Block, Element, getSimpleAttributeName, Visitor } from './ast';

export const LEADING_TRIVIA_CHARS = [' ', '\n', '\r', '\t'];

export type ParseTemplateResult = {
    tree?: { rootNodes: unknown };
    errors: ParseTemplateError[];
};

export type ParseTemplateError = { type: string; error: ParseError };

export interface ParseError {
    msg: string;
    contextualMessage(): string;
    toString(): string;
}

export type TransformTemplateAttributesResult = {
    fileContent: string;
    changed?: boolean;
    errors: ParseTemplateError[];
};

/**
 * Represents a range of text within a file. Omitting the end
 * means that it's until the end of the file.
 */
export type Range = {
    start: number;
    end?: number;
    node: ts.Node;
    type: string;
    remove: boolean;
};

export type Replacement = { from: string; to: string };
export type MigrationData = {
    elementName: string;
    attrs: { key: Replacement; value: { replacements: Replacement[]; default: string } };
};

export function forEachClass(sourceFile: ts.SourceFile, callback: (node: ts.ClassDeclaration) => void) {
    sourceFile.forEachChild(function walk(node) {
        if (ts.isClassDeclaration(node)) {
            callback(node);
        }
        node.forEachChild(walk);
    });
}

export function canMigrateFile(sourceFile: ts.SourceFile, program: ts.Program): boolean {
    // We shouldn't migrate .d.ts files, files from an external library or type checking files.
    return !(
        sourceFile.fileName.endsWith('.ngtypecheck.ts') ||
        sourceFile.isDeclarationFile ||
        program.isSourceFileFromExternalLibrary(sourceFile)
    );
}

export function updateDecoratorProperty(decorator: ts.Decorator, propertyName: string, newValue: string): ts.Decorator {
    if (ts.isCallExpression(decorator.expression)) {
        const [arg] = decorator.expression.arguments;

        if (ts.isObjectLiteralExpression(arg)) {
            const newProperties = arg.properties.map((prop) => {
                if (ts.isPropertyAssignment(prop) && ts.isIdentifier(prop.name) && prop.name.text === propertyName) {
                    return ts.factory.updatePropertyAssignment(
                        prop,
                        prop.name,
                        ts.factory.createNoSubstitutionTemplateLiteral(newValue, newValue)
                    );
                }
                return prop;
            });

            const newArg = ts.factory.updateObjectLiteralExpression(arg, newProperties);
            const newCall = ts.factory.updateCallExpression(
                decorator.expression,
                decorator.expression.expression,
                undefined,
                [newArg]
            );

            return ts.factory.updateDecorator(decorator, newCall);
        }
    }
    return decorator;
}

export function getClassWithUpdatedDecorator(
    node: ts.ClassDeclaration,
    decorator: ts.Decorator,
    updatedDecorator: ts.Decorator
) {
    const nodeDecorators = ts.getDecorators(node) || [];
    const decoratorIndex = nodeDecorators.indexOf(decorator);
    const updatedDecorators = decoratorIndex !== -1 ? [
                  ...nodeDecorators.slice(0, decoratorIndex),
                  updatedDecorator,
                  ...nodeDecorators.slice(decoratorIndex + 1, nodeDecorators.length)] : nodeDecorators;
    const modifiers = node.modifiers ? node.modifiers.filter((modifier) => modifier !== decorator) : [];

    return ts.factory.replaceDecoratorsAndModifiers(node, [...updatedDecorators, ...modifiers]);
}

/**
 * Class that collects specific elements for migration
 * based on matching tag names and attribute values.
 *
 * Used as reference:
 * @see https://github.com/angular/angular/blob/6fa8d441979fdbabb88dddd246f54587e17126e8/packages/core/schematics/migrations/control-flow-migration/types.ts#L432
 */
export class ElementCollector implements Visitor {
    readonly elementsToMigrate: Element[] = [];

    constructor(
        public migrationData: MigrationData,
        public filePath?: string
    ) {}

    visitElement(el: Element): void {
        if (el.name === this.migrationData.elementName && el.attrs.length > 0) {
            for (const attr of el.attrs) {
                const cleanAttrName = getSimpleAttributeName(attr.name);
                if (cleanAttrName === this.migrationData.attrs.key.from) {
                    this.elementsToMigrate.push(el);
                }
            }
        }
        this.visitChildren(el);
    }

    /**
     * Recursively visits all child nodes of a given element or block.
     */
    visitChildren(el: Element | Block) {
        if (el.children?.length) {
            for (const child of el.children) {
                child.visit(this);
            }
        }
    }

    /**
     * Handle @if, @for blocks
     */
    visitBlock(block: Block, _1: any) {
        this.visitChildren(block);
    }

    // part of interface
    visitAttribute(_: Attribute, _1: any) {}
    visitText(_: any, _1: any) {}
    visitComment(_: any, _1: any) {}
    visitExpansion(_: any, _1: any) {}
    visitExpansionCase(_: any, _1: any) {}
    visitBlockParameter(_: any, _1: any) {}
    visitLetDeclaration(_: any, _1: any) {}
}

export class AnalyzedFile {
    private ranges: Range[] = [];
    sourceFile: ts.SourceFile;
    templateRanges: Range[] = [];

    constructor(sourceFile: ts.SourceFile) {
        this.sourceFile = sourceFile;
    }

    /** Returns the ranges in the order in which they should be migrated. */
    getSortedRanges(): Range[] {
        // templates first for checking on whether certain imports can be safely removed
        this.templateRanges = this.ranges
            .slice()
            .filter((x) => x.type === 'template' || x.type === 'templateUrl')
            .sort((aStart, bStart) => bStart.start - aStart.start);
        return this.templateRanges;
    }

    /**
     * Adds a text range to an `AnalyzedFile`.
     * @param path Path of the file.
     * @param sourceFile SourceFile
     * @param analyzedFiles Map keeping track of all the analyzed files.
     * @param range Range to be added.
     */
    static addRange(
        path: string,
        sourceFile: ts.SourceFile,
        analyzedFiles: Map<string, AnalyzedFile>,
        range: Range
    ): void {
        let analysis = analyzedFiles.get(path);

        if (!analysis) {
            analysis = new AnalyzedFile(sourceFile);
            analyzedFiles.set(path, analysis);
        }

        const duplicate = analysis.ranges.find((current) => current.start === range.start && current.end === range.end);

        if (!duplicate) {
            analysis.ranges.push(range);
        }
    }
}

/**
 * parses the template string into the Html AST
 */
export async function parseTemplate(template: string): Promise<ParseTemplateResult> {
    try {
        const { HtmlParser } = await import('@angular/compiler');
        const parser = new HtmlParser();

        // Note: we use the HtmlParser here, instead of the `parseTemplate` function, because the
        // latter returns an Ivy AST, not an HTML AST. The HTML AST has the advantage of preserving
        // interpolated text as text nodes containing a mixture of interpolation tokens and text tokens,
        // rather than turning them into `BoundText` nodes like the Ivy AST does. This allows us to
        // easily get the text-only ranges without having to reconstruct the original text.
        const parsed = parser.parse(template, '', {
            // Allows for ICUs to be parsed.
            tokenizeExpansionForms: true,
            // Explicitly disable blocks so that their characters are treated as plain text.
            tokenizeBlocks: true,
            preserveLineEndings: true,
            leadingTriviaChars: LEADING_TRIVIA_CHARS
        });

        // Don't migrate invalid templates.
        if (parsed.errors && parsed.errors.length > 0) {
            const errors = parsed.errors.map((e) => ({ type: 'parse', error: e }));
            return { tree: undefined, errors };
        }
        return { tree: parsed, errors: [] };
    } catch (e: any) {
        return { tree: undefined, errors: [{ type: 'parse', error: e }] };
    }
}
