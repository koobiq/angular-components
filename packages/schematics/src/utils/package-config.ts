import { SchematicsException, Tree } from '@angular-devkit/schematics';
import { ProjectDefinition, readWorkspace } from '@schematics/angular/utility';
import * as messages from './messages';

/**
 * Sorts the keys of the given object.
 * @returns A new object instance with sorted keys
 */
function sortObjectByKeys(obj: Record<string, string>) {
    return Object.keys(obj)
        .sort()
        .reduce((result, key) => (result[key] = obj[key]) && result, {});
}

/**
 * Adds a package to the package.json in the given tree
 */
export function addPackageToPackageJson(tree: Tree, pkg: string, version: string, devDependency = false): Tree {
    if (tree.exists('package.json')) {
        const sourceText = tree.read('package.json')!.toString('utf-8');
        const json = JSON.parse(sourceText);

        const dependenciesType = devDependency ? 'devDependencies' : 'dependencies';
        const dependencies = json[dependenciesType] || {};

        if (!dependencies[pkg]) {
            dependencies[pkg] = version;
            json[dependenciesType] = sortObjectByKeys(dependencies);
        }

        tree.overwrite('package.json', JSON.stringify(json, null, 2));
    }

    return tree;
}

/**
 * Gets the version of the specified package by looking at the package.json in the given tree
 */
export function getPackageVersionFromPackageJson(
    tree: Tree,
    name: string,
    includeDevDependencies = false
): string | null {
    if (!tree.exists('package.json')) {
        return null;
    }

    const packageJson = JSON.parse(tree.read('package.json')!.toString('utf8'));

    if (packageJson.dependencies && packageJson.dependencies[name]) {
        return packageJson.dependencies[name];
    }

    if (includeDevDependencies && packageJson.devDependencies && packageJson.devDependencies[name]) {
        return packageJson.devDependencies[name];
    }

    return null;
}

export async function setupOptions(
    project: string | undefined,
    tree: Tree
): Promise<ProjectDefinition | undefined | never> {
    if (project) {
        const workspace = await readWorkspace(tree);
        const projectWorkspace = workspace.projects.get(project);

        if (!projectWorkspace) {
            throw new SchematicsException(messages.noProject(project));
        }

        return projectWorkspace;
    }
}

export async function getParsingInfo(project: string | undefined, tree: Tree) {
    const tsPaths = new Set<string>();
    const templatePaths = new Set<string>();
    const projectDefinition = await setupOptions(project, tree);
    if (!projectDefinition) {
        throw new SchematicsException(messages.noProject('no project'));
    }

    tree.getDir(projectDefinition.root).visit((filePath: string) => {
        if (filePath.endsWith('.ts')) {
            tsPaths.add(filePath);
        }

        if (filePath.endsWith('.html')) {
            templatePaths.add(filePath);
        }
    });

    return { tsPaths, templatePaths, projectDefinition };
}

/**
 * WARNING: do not change directly, since it's a copy of
 * https://github.com/angular/angular/blob/6fa8d441979fdbabb88dddd246f54587e17126e8/packages/core/schematics/utils/load_esm.ts
 *
 * This uses a dynamic import to load a module which may be ESM.
 * CommonJS code can load ESM code via a dynamic import. Unfortunately, TypeScript
 * will currently, unconditionally downlevel dynamic import into a require call.
 * require calls cannot load ESM code and will result in a runtime error. To workaround
 * this, a Function constructor is used to prevent TypeScript from changing the dynamic import.
 * Once TypeScript provides support for keeping the dynamic import this workaround can
 * be dropped.
 * This is only intended to be used with Angular framework packages.
 *
 * @param modulePath The path of the module to load.
 * @returns A Promise that resolves to the dynamically imported module.
 */
export async function loadEsmModule<T>(modulePath: string | URL): Promise<T> {
    const namespaceObject = await new Function('modulePath', `return import(modulePath);`)(modulePath);

    // If it is not ESM then the values needed will be stored in the `default` property.
    // TODO_ESM: This can be removed once `@angular/*` packages are ESM only.
    if (namespaceObject.default) {
        return namespaceObject.default;
    } else {
        return namespaceObject;
    }
}
