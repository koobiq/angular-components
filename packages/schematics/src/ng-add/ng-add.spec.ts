import { SchematicTestRunner, UnitTestTree } from '@angular-devkit/schematics/testing';
import { join } from 'path';
import { createTestApp } from '../utils/testing';

const getPackageJsonDependencies = (tree: UnitTestTree) => {
    return JSON.parse(tree.get('/package.json')!.content.toString()).dependencies;
};

describe(`ng add '@koobiq/components'`, () => {
    let runner: SchematicTestRunner;
    let appTree: UnitTestTree;

    beforeEach(async () => {
        runner = new SchematicTestRunner('schematics', join(__dirname, '../collection.json'));
        appTree = await createTestApp(runner);
    });

    it(`should add missing dependencies to 'package.json'`, async () => {
        expect(getPackageJsonDependencies(appTree)).toMatchSnapshot('before running schematics');
        const tree = await runner.runSchematic('ng-add', {}, appTree);
        expect(getPackageJsonDependencies(tree)).toMatchSnapshot('after running schematics');
    });

    it(`should report when specified 'project' is not found`, async () => {
        await expect(runner.runSchematic('ng-add', { project: 'test' }, appTree)).rejects.toThrow(
            "Unable to find project 'test' in the workspace"
        );
    });
});
