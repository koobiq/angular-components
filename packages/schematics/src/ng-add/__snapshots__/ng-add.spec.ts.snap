// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ng add '@koobiq/components' should add missing dependencies to 'package.json': after running schematics 1`] = `
{
  "@angular/animations": "^18.2.0",
  "@angular/cdk": "^0.0.0",
  "@angular/common": "^18.2.0",
  "@angular/compiler": "^18.2.0",
  "@angular/core": "^18.2.0",
  "@angular/forms": "^18.2.0",
  "@angular/platform-browser": "^18.2.0",
  "@angular/platform-browser-dynamic": "^18.2.0",
  "@angular/router": "^18.2.0",
  "@koobiq/angular-luxon-adapter": "^0.0.0",
  "@koobiq/cdk": "^0.0.0",
  "@koobiq/date-adapter": "^0.0.0",
  "@koobiq/date-formatter": "^0.0.0",
  "@koobiq/design-tokens": "^0.0.0",
  "@koobiq/icons": "^0.0.0",
  "@koobiq/tokens-builder": "^0.0.0",
  "@messageformat/core": "^0.0.0",
  "@radix-ng/primitives": "^0.0.0",
  "luxon": "^0.0.0",
  "rxjs": "~7.8.0",
  "tslib": "^2.3.0",
  "zone.js": "~0.14.10",
}
`;

exports[`ng add '@koobiq/components' should add missing dependencies to 'package.json': before running schematics 1`] = `
{
  "@angular/animations": "^18.2.0",
  "@angular/common": "^18.2.0",
  "@angular/compiler": "^18.2.0",
  "@angular/core": "^18.2.0",
  "@angular/forms": "^18.2.0",
  "@angular/platform-browser": "^18.2.0",
  "@angular/platform-browser-dynamic": "^18.2.0",
  "@angular/router": "^18.2.0",
  "rxjs": "~7.8.0",
  "tslib": "^2.3.0",
  "zone.js": "~0.14.10",
}
`;
